apps/sms/test/marionette/generators/thread.js

```diff
@@ -5,48 +5,116 @@
+};
+
+var DEFAULT_SMS = _.assign({
+  type: 'sms',
+  body: 'Thread message content',
+  receiver: null,
+  messageClass: 'class-1',
+  deliveryStatus: 'not-applicable',
+  deliveryTimestamp: 0
+}, DEFAULT_MESSAGE);
+
+var DEFAULT_MMS = _.assign({
+  type: 'mms',
+  receivers: [],
+  deliveryInfo: [{ deliveryStatus: 'not-applicable' }],
+  subject: 'Message subject',
+  smil: null,
+  attachments: [],
+  expiryDate: 0,
+  readReportRequested: false
+}, DEFAULT_MESSAGE);
+
 var ThreadGenerator = {
   uniqueThreadId: 0,
   uniqueMessageId: 0,
 
   generate: function(parameters) {
-
-    var DEFAULT_PARAMETERS = {
-      numberOfMessages: 1,
-      messageType: 'sms',
-      participants: ['+123'],
-      body: 'Thread message content'
-    };
-
-    parameters = _.assign(DEFAULT_PARAMETERS, parameters);
+    parameters = _.assign({}, DEFAULT_THREAD_PARAMETERS, parameters);
 
     var thread = {
       id: ++ThreadGenerator.uniqueThreadId,
       body: parameters.body,
       lastMessageType: parameters.messageType,
-      timestamp: Date.now(),
+      timestamp: 0,
       messages: [],
       participants: parameters.participants
     };
 
     for (var i = 0; i < parameters.numberOfMessages; i++) {
-      var id = ++ThreadGenerator.uniqueMessageId;
-      thread.messages.push({
-        id: id,
-        iccId: null,
-        threadId: thread.id,
-        sender: null,
-        receiver: parameters.participants[0],
-        type: parameters.messageType,
-        delivery: 'sent',
-        body: thread.body + ' (messageId: ' + id + ')',
-        timestamp: Date.now()
-      });
+      var message = null;
+      var messageBody = thread.body + ' (messageId: ' +
+        (ThreadGenerator.uniqueMessageId + 1) + ')';
+      if (parameters.messageType === 'sms') {
+        message = this.generateSMS({
+          threadId: thread.id,
+          receiver:  parameters.participants[0],
+          delivery: parameters.delivery,
+          body: messageBody,
+          timestamp: Date.now() - i * 10000
+        });
+      } else {
+        message = this.generateMMS({
+          threadId: thread.id,
+          receivers: parameters.participants,
+          delivery: parameters.delivery,
+          expiryDate: parameters.expiryDate,
+          timestamp: Date.now() - i * 10000,
+          subject: parameters.subject,
+          attachments: parameters.attachments || [{
+            type: 'text/plain', content: messageBody
+          }]
+        });
+      }
+
+      thread.messages.push(message);
     }
 
+    thread.timestamp =
+      thread.messages[parameters.numberOfMessages - 1].timestamp;
+
     return thread;
-  }
+  },
+
+  generateSMS: function(smsParameters) {
+    if (!smsParameters.threadId) {
+      throw new Error('ThreadId should be specified!');
+    }
 
+    return _.assign(
+      { id: smsParameters.id || ++ThreadGenerator.uniqueMessageId },
+      DEFAULT_SMS,
+      smsParameters
+    );
+  },
+
+  generateMMS: function(mmsParameters) {
```

Comment:
Using these public methods we can create any threads we want without pretty dumb ```generate``` function.